Libraries/JavaScriptAppEngine/System/JSTimers/JSTimersExecution.js

```diff
@@ -108,6 +117,34 @@ var JSTimersExecution = {
     }
   },
 
+  callIdleCallbacks: function(frameTime) {
+    // Arbitrary threshold: 10ms remaining in frame.
+    if (17 - (performanceNow() - frameTime) < 10) {
```

Comment:
Note that scheduling running into the next frame is the _only_ reason this API exists at all in browsers. If you wanted to ensure to have ample time at the end of a frame you can just use `setTimeout`.

If you assume that you take close to entire frame to update a frame, you are already in very dangerous territory because you're likely to run across - because of a miscalculation, slower device or OS scheduling interrupting you. So the argument goes two ways. If your idle time is too risky at the edges then so is the frame calculation.

In practice you don't need that much time to calculate the next frame, outside of compositing. This heuristic seems to assume that it normally takes less than 6ms. If that's the case then there is plenty of headroom even if you run over.

This is the only reason this API is competitive with threads.

Btw, Chrome by default gives you 50ms before any animation starts. Because dropping a frame on the initial interaction is NBD. When there are animations happening it cuts down but still leaving you ample time to utilize as many cycles as possible.